apiserver/common/upgradeseries.go

```diff
@@ -152,6 +152,44 @@ func (u *UpgradeSeriesAPI) SetUpgradeSeriesStatus(args params.SetUpgradeSeriesSt
 	return u.setUpgradeSeriesStatus(args)
 }
 
+func (u *UpgradeSeriesAPI) CompleteUnitUpgradeSeries(args params.SetUpgradeSeriesStatusParams) (params.ErrorResults, error) {
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Params)),
+	}
+	canAccess, err := u.accessUnit()
+	if err != nil {
+		return params.ErrorResults{}, err
+	}
+	for i, p := range args.Params {
+		//TODO[externalreality] refactor all of this, its being copied often.
+		tag, err := names.ParseUnitTag(p.Entity.Tag)
+		if err != nil {
+			result.Results[i].Error = ServerError(ErrPerm)
+			continue
+		}
+		if !canAccess(tag) {
+			result.Results[i].Error = ServerError(ErrPerm)
+			continue
+		}
+		unit, err := u.getUnit(tag)
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+		status, err := model.ValidateUnitSeriesUpgradeStatus(p.Status)
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+		err = unit.SetUpgradeSeriesStatus(status, statusType)
```

Comment:
I'm guessing this doesn't compile, as I don't see where statusType comes from. But maybe it isn't a big deal as later patches will remove this parameter?